burstphoto/align.metal

```diff
@@ -416,3 +740,992 @@ kernel void compute_merge_weight(texture2d<float, access::read> texture_diff [[t
+                norm_cosine1 = (0.5f-cosine_factor*cos(-angle*(dm+1.5f)))*(0.5f-cosine_factor*cos(-angle*(dy+0.5f)));
+                         
+                // calculate coefficients
+                coefRe = cos(angle*dn*dy);
+                coefIm = sin(angle*dn*dy);
+      
+                dataRe = norm_cosine0*tmp_data[2*dy+0];
+                Re0 += (coefRe * dataRe);
+                Im0 += (coefIm * dataRe);
+                
+                dataRe = norm_cosine1*tmp_data[2*dy+1];
+                Re1 += (coefRe * dataRe);
+                Im1 += (coefIm * dataRe);
+            }
+            
+            // write into temporary textures
+            tmp_texture_ft.write(Re0, uint2(2*m+0, n));
+            tmp_texture_ft.write(Im0, uint2(2*m+1, n));
+            tmp_texture_ft.write(Re1, uint2(2*m+2, n));
+            tmp_texture_ft.write(Im1, uint2(2*m+3, n));
+        }
+    }
+        
+    // row-wise one-dimensional fast Fourier transform along x-direction
+    // exploit symmetry of real dft and calculate reduced number of rows
+    for (int dn = 0; dn <= tile_size/2; dn++) {
+        
+        int const n = n0 + dn;        
+        
+        // copy data to temp vector
+        for (int dm = 0; dm < tile_size; dm++) {
+            tmp_data[2*dm+0] = tmp_texture_ft.read(uint2(2*(m0+dm)+0, n));
+            tmp_data[2*dm+1] = tmp_texture_ft.read(uint2(2*(m0+dm)+1, n));
+        }
+        
+        // calculate 4 small discrete Fourier transforms
+        for (int dm = 0; dm < tile_size/4; dm++) {
+             
+            int const m = 2*(m0 + dm);
+            
+            // fill with zeros
+            Re0 = Im0 = Re1 = Im1 = Re2 = Im2 = Re3 = Im3 = zeros;
+            
+            for (int dx = 0; dx < tile_size; dx+=4) {
+              
+                // calculate coefficients
+                coefRe = cos(angle*dm*dx);
+                coefIm = sin(angle*dm*dx);
+                                
+                // DFT0
+                dataRe = tmp_data[2*dx+0];
+                dataIm = tmp_data[2*dx+1];
+                Re0   += (coefRe*dataRe - coefIm*dataIm);
+                Im0   += (coefIm*dataRe + coefRe*dataIm);
+                // DFT1
+                dataRe = tmp_data[2*dx+2];
+                dataIm = tmp_data[2*dx+3];
+                Re2   += (coefRe*dataRe - coefIm*dataIm);
+                Im2   += (coefIm*dataRe + coefRe*dataIm);
+                // DFT2
+                dataRe = tmp_data[2*dx+4];
+                dataIm = tmp_data[2*dx+5];
+                Re1   += (coefRe*dataRe - coefIm*dataIm);
+                Im1   += (coefIm*dataRe + coefRe*dataIm);
+                // DFT3
+                dataRe = tmp_data[2*dx+6];
+                dataIm = tmp_data[2*dx+7];
+                Re3   += (coefRe*dataRe - coefIm*dataIm);
+                Im3   += (coefIm*dataRe + coefRe*dataIm);
+            }
+            
+            // first butterfly to combine results
```

Comment:
butterfly?

Reply:
The diagrams of a Fast Fourier Transform include many cross-combinations of values. These are sometimes called "butterfly" as they may be associated with a butterfly by some people :-)